### PR TITLE
Implement rfc0035: default `name` from `pname`

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change

Implements https://github.com/NixOS/rfcs/pull/35

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by shlevy — APPROVED** [review]

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -78,6 +82,10 @@ rec {
 
     , ... } @ attrs:
 
+    # Check that the name is consistent with pname and version:
+    assert lib.lists.all (name: builtins.hasAttr name attrs) ["name" "pname" "version"]
+      -> lib.strings.hasSuffix "${attrs.pname}-${attrs.version}" attrs.name;
```
Comment (Profpatsch) [inline pkgs/stdenv/generic/make-derivation.nix]: Are you sure this is the way to go?

I think not allowing both `pname` & `version` and `name` would lead to less confusion.

Also, use `lib.assertMsg` for this to add an error message.

**Review by Profpatsch — CHANGES_REQUESTED** [review]

File: pkgs/stdenv/generic/make-derivation.nix (hunk 1 above)
Comment (Synthetica9) [reply pkgs/stdenv/generic/make-derivation.nix]: > I think not allowing both `pname` & `version` and `name` would lead to less confusion.

I (mostly) agree. However, that would require tree-wide changes to packages already using the following pattern:

```nix
mkDerivation rec {
  pname = "foo";
  version = "bar";
  name = "${pname}-${version}"

  # ...
}
``` 

Of which there are quite a few.

The follow-up to this PR is of course a series of PR's eliminating this pattern, but I don't think it should all be done in one PR (I tried to do that in the previous PR, and it went stale rather quickly, so I'd like to keep this PR relatively small).

After this all is fixed, I'd say we shouldn't even allow `pname` and `name`. 

> Also, use `lib.assertMsg` for this to add an error message.

Ah, yes, of course, fixing that!

File: pkgs/stdenv/generic/make-derivation.nix [hunk 2]
```diff
@@ -78,6 +82,12 @@ rec {
 
     , ... } @ attrs:
 
+    # Check that the name is consistent with pname and version:
+    assert lib.assertMsg
+      (lib.lists.all (name: builtins.hasAttr name attrs) ["name" "pname" "version"]
+        -> lib.strings.hasSuffix "${attrs.pname}-${attrs.version}" attrs.name)
+      "mkDerivation: `name` must be consistent with `pname-version`";
```
Comment (Profpatsch) [inline pkgs/stdenv/generic/make-derivation.nix]: I’d always include what the actual values are, otherwise they get lost in the stack traces. So in this case:

```
"mkDerivation: `name` ("${attrs.name}") must be consistent with `pname-version` ("${attrs.pname}-${attrs.version}").";
```
(untested)

Though I still think the check should disallow having both `name` and `pname`/`version`.

**Review by Profpatsch — CHANGES_REQUESTED** [review]

File: pkgs/stdenv/generic/make-derivation.nix (hunk 2 above)
Comment (Synthetica9) [reply pkgs/stdenv/generic/make-derivation.nix]: Solved in 1f7fc09

**Review by Profpatsch — APPROVED** [review]

**Comment by dtzWill** [comment]
Is someone working on the fallout this causes re:pname/name violations? Surprised that wasn't done as part of the PR or explicitly discussed at least....

Anyway happily it seems to be just one package blocking my eval so no big deal :).
(modemmanager)

**Comment by danieldk** [comment]
So far, I have bumped into modemmanager and NetworkManager-l2tp.

**Comment by c0bw3b** [comment]
Just trying to nox-review libmtp I get all this......
https://gist.github.com/c0bw3b/64ce22facbb41144761a3e7f29d02fc1

Also if certain use of `pname` and `name` become illegal then it should be documented before being enforced.

**Comment by FRidh** [comment]
This is causing evaluation errors.

Before enforcing something like this, make sure you fix all occasions.

**Comment by FRidh** [comment]
Reverted in abea6f461ae886544917e32634e5612b0fc73c13.

**Comment by Synthetica9** [comment]
I have implemented a different check, that doesn't seem to have any false positive:

```
      1 trace: pname Metaphone does not appear in name python2.7-metaphone-0.6
      1 trace: pname sat-0.6.1 does not appear in name salut-a-toi-0.6.1
      1 trace: version 0.0.0 does not appear in name python2.7-waitress-django
      1 trace: version 0.0.0 does not appear in name python3.7-waitress-django
      1 trace: version 0.0.17 does not appear in name typeshed
      1 trace: version 0.11.1 does not appear in name mist
      1 trace: version 0.1 does not appear in name python2.7-Snapper-GUI
      1 trace: version 0.1 does not appear in name python3.7-Snapper-GUI
      1 trace: version 0.25.2 does not appear in name brave
      1 trace: version 0.5.0 does not appear in name icr
      1 trace: version 0.5.5 does not appear in name python3.7-pathspec-0.5.9
      1 trace: version 0.6.1 does not appear in name kubectx
      1 trace: version 0.7.1.20180919 does not appear in name scry
      1 trace: version 0.7.2 does not appear in name python2.7-rbtools
      1 trace: version 0.7.3 does not appear in name vimiv
      1 trace: version 0.9.0 does not appear in name cadence
      1 trace: version 1.01 does not appear in name hid-listen
      1 trace: version 1.0b10 does not appear in name python2.7-html5lib-1.0.1
      1 trace: version 1.0 does not appear in name arm-trusted-firmware-sun50iw1p1-1.5
      1 trace: version 1.0 does not appear in name python3.7-urwidtrees
      1 trace: version 1.13.0 does not appear in name mupdf-1.10a
      1 trace: version 1.1 does not appear in name yojimbo
      1 trace: version 1.2.32 does not appear in name remmina-19.03pre-git
      1 trace: version 1.2.sakura.5 does not appear in name oneko-1.2.5
      1 trace: version 1.5 does not appear in name chicken-aes-export
      1 trace: version 1.7.9 does not appear in name python2.7-sphinx-1.2.3
      1 trace: version 2015-08-05 does not appear in name python2.7-repocheck
      1 trace: version 2015-08-05 does not appear in name python3.7-repocheck
      1 trace: version 2016-10-18 does not appear in name python3.7-pyinputevent
      1 trace: version 2017-06-04 does not appear in name beetle-pce-fast
      1 trace: version 2017-06-04 does not appear in name beetle-psx
      1 trace: version 2017-06-04 does not appear in name beetle-saturn
      1 trace: version 20180606 does not appear in name http-getter
```

I think I can have all those fixed in a few hours, I'll open a new PR then.

**Comment by dtzWill** [comment]
Awesome!! I'm excited for this-- looking forward to taking advantage of it.  The checks were unexpected from the title, but huge +1 for working to clean up all out expressions and helping them stay that way.
Thanks for tackling this!! Look forward to the next PR, hopefully the fixes are straightforward and not too tedious.

FWIW I only had one or two additional packages to fix, and I think they've been mentioned or addressed already. (which wasn't expected when I grumpily wrote my earlier comment)

In conclusion thank you and +1. :)

**Comment by Synthetica9** [comment]
@dtzWill It mostly seems to be a case of people mistaking `name` for what `pname` does after this PR

**Comment by samueldr** [comment]
@Synthetica9, or others involved with the change: can we have documentation in the Nixpkgs manual for `pname`? It would be great to have documentation available for this change to point at in PR reviews, regarding the usage of `pname`.

I'm thinking mainly of details like (if there are any cases) when not to use `pname`, and other important details to know.
